### PR TITLE
[MXNET-1194] Reenable nightly tutorials tests for Python2 and Python3

### PR DESCRIPTION
**Pull request by ThomasDelteil**
## Description ##

This is really overdue. Tutorial tests were disabled because of flakiness and cost concerns. Flakiness was addressed but the tutorials were never reenabled. This PR reenable them, and fix the tutorials that were not passing anymore. There is one known tutorial not passing the test, it is the ONNX export one but that's a bug introduced yesterday in master. @Roshrini is aware and has a fix for it.

### Essentials ###
Please feel free to remove inapplicable items for your PR.
- [x] The PR title starts with [MXNET-$JIRA_ID], where $JIRA_ID refers to the relevant [JIRA issue](https://issues.apache.org/jira/projects/MXNET/issues) created (except PRs with tiny changes)
- [x] Changes are complete (i.e. I finished coding on this PR)
- [x] All changes have test coverage:
- [x] Code is well-documented: 
- [x] To the my best knowledge, examples are either not affected by this change, or have been fixed to be compatible with this change

### Changes ###
- [x] Added a new settings in make docs to only build the doxygen docs that create the notebooks

FYI @marcoabreu @aaronmarkham @vishaalkapoor @Roshrini 

Tests are passing:
```
root@359b44f21040:/work/mxnet/tests/tutorials# nosetests-3.4 $NOSE_COVERAGE_ARGUMENTS --with-xunit --xunit-file nosetests_tutorials.xml test_tutorials.py --nologcapture
....[21:31:40] src/io/iter_image_recordio_2.cc:172: ImageRecordIOParser2: ./data/caltech.rec, use 4 threads for decoding..
.......[21:34:03] src/operator/nn/./cudnn/./cudnn_algoreg-inl.h:97: Running performance tests to find the best convolution algorithm, this can take a while... (setting env variable MXNET_CUDNN_AUTOTUNE_DEFAULT to 0 to disable)
...[21:34:43] src/operator/nn/./cudnn/./cudnn_algoreg-inl.h:97: Running performance tests to find the best convolution algorithm, this can take a while... (setting env variable MXNET_CUDNN_AUTOTUNE_DEFAULT to 0 to disable)
...[21:35:20] src/operator/nn/./cudnn/./cudnn_algoreg-inl.h:97: Running performance tests to find the best convolution algorithm, this can take a while... (setting env variable MXNET_CUDNN_AUTOTUNE_DEFAULT to 0 to disable)
.[21:36:09] src/operator/nn/./cudnn/./cudnn_algoreg-inl.h:97: Running performance tests to find the best convolution algorithm, this can take a while... (setting env variable MXNET_CUDNN_AUTOTUNE_DEFAULT to 0 to disable)
..[21:37:43] src/operator/nn/./cudnn/./cudnn_algoreg-inl.h:97: Running performance tests to find the best convolution algorithm, this can take a while... (setting env variable MXNET_CUDNN_AUTOTUNE_DEFAULT to 0 to disable)
..[21:39:08] src/nnvm/legacy_json_util.cc:209: Loading symbol saved by previous version v0.8.0. Attempting to upgrade...
[21:39:08] src/nnvm/legacy_json_util.cc:217: Symbol successfully upgraded!
.[21:39:54] src/operator/nn/./cudnn/./cudnn_algoreg-inl.h:97: Running performance tests to find the best convolution algorithm, this can take a while... (setting env variable MXNET_CUDNN_AUTOTUNE_DEFAULT to 0 to disable)
[21:42:59] src/operator/nn/./cudnn/./cudnn_algoreg-inl.h:97: Running performance tests to find the best convolution algorithm, this can take a while... (setting env variable MXNET_CUDNN_AUTOTUNE_DEFAULT to 0 to disable)
......[21:44:59] src/operator/nn/./cudnn/./cudnn_algoreg-inl.h:97: Running performance tests to find the best convolution algorithm, this can take a while... (setting env variable MXNET_CUDNN_AUTOTUNE_DEFAULT to 0 to disable)
.[21:45:14] src/nnvm/legacy_json_util.cc:209: Loading symbol saved by previous version v0.8.0. Attempting to upgrade...
[21:45:14] src/nnvm/legacy_json_util.cc:217: Symbol successfully upgraded!
.....[21:45:54] src/operator/nn/./cudnn/./cudnn_algoreg-inl.h:97: Running performance tests to find the best convolution algorithm, this can take a while... (setting env variable MXNET_CUDNN_AUTOTUNE_DEFAULT to 0 to disable)
.[21:45:58] src/operator/././../common/utils.h:450: Optimizer with lazy_update = True detected. Be aware that lazy update with row_sparse gradient is different from standard update, and may lead to different empirical results. See https://mxnet.incubator.apache.org/api/python/optimization/optimization.html for more details.
..[21:46:11] src/operator/././../common/utils.h:450: Optimizer with lazy_update = True detected. Be aware that lazy update with row_sparse gradient is different from standard update, and may lead to different empirical results. See https://mxnet.incubator.apache.org/api/python/optimization/optimization.html for more details.
..[21:46:41] src/operator/nn/./cudnn/./cudnn_algoreg-inl.h:97: Running performance tests to find the best convolution algorithm, this can take a while... (setting env variable MXNET_CUDNN_AUTOTUNE_DEFAULT to 0 to disable)
[21:46:41] src/operator/nn/./cudnn/./cudnn_algoreg-inl.h:97: Running performance tests to find the best convolution algorithm, this can take a while... (setting env variable MXNET_CUDNN_AUTOTUNE_DEFAULT to 0 to disable)
....
----------------------------------------------------------------------
Ran 44 tests in 1066.832s

OK
```

### CODE REVIEW

File: docs/tutorials/gluon/learning_rate_finder.md [hunk 1]
```diff
@@ -109,7 +108,7 @@ class ContinuousBatchSampler():
 
 sampler = mx.gluon.data.RandomSampler(len(dataset))
 batch_sampler = ContinuousBatchSampler(sampler, batch_size=128)
-data_loader = mx.gluon.data.DataLoader(dataset, batch_sampler=batch_sampler, num_workers=cpu_count())
+data_loader = mx.gluon.data.DataLoader(dataset, batch_sampler=batch_sampler)
```
[the remark this answers is not in the document]
Comment (vishaalkapoor) [reply docs/tutorials/gluon/learning_rate_finder.md]: And if you don't remove parallel data loading, should the num_workers be the max of 1 and cpu_count() - 3 as a rule of thumb?

File: docs/tutorials/gluon/learning_rate_schedules.md [hunk 2]
```diff
@@ -134,7 +134,7 @@ batch_size = 64
 
 # Load the training data
 train_dataset = mx.gluon.data.vision.MNIST(train=True).transform_first(transforms.ToTensor())
-train_dataloader = mx.gluon.data.DataLoader(train_dataset, batch_size, shuffle=True)
+train_dataloader = mx.gluon.data.DataLoader(train_dataset, batch_size, shuffle=True, num_workers=5)
```
Comment (vishaalkapoor) [inline docs/tutorials/gluon/learning_rate_schedules.md]: how come num_workers is hard-coded to 5 instead of using max(1,cpu_count-3)?

Comment (vishaalkapoor) [reply docs/tutorials/gluon/learning_rate_schedules.md]: Are you removing all python's in ```python? There's one on line 124.

File: docs/tutorials/onnx/export_mxnet_to_onnx.md [hunk 3]
```diff
@@ -45,11 +45,11 @@ Now, we have downloaded ResNet-18 symbol, params and synset file on the disk.
 
 Let us describe the MXNet's `export_model` API. 
 
-```python
+```
```
[the remark this answers is not in the document]
Comment (vishaalkapoor) [reply docs/tutorials/onnx/export_mxnet_to_onnx.md]: +1, is default code-highlighting working just fine?

File: docs/tutorials/onnx/fine_tuning_gluon.md [hunk 4]
```diff
@@ -122,7 +122,7 @@ We need to transform the images to a format accepted by the network
 EDGE = 224
 SIZE = (EDGE, EDGE)
 BATCH_SIZE = 32
-NUM_WORKERS = multiprocessing.cpu_count()
+NUM_WORKERS = max(multiprocessing.cpu_count()-6, 0)
```
[the remark this answers is not in the document]
Comment (vishaalkapoor) [reply docs/tutorials/onnx/fine_tuning_gluon.md]: there should be a max(1, f(cpu_count)), but f(n) is usu n - 3 as mentioned above. Why n - 6?

**Comment by marcoabreu** [comment]
After the initial review, we should set up a test job to validate the changes in this Jenkinsfile to ensure everything works as expected. Thanks a lot for your help to get the tutorials back up again!

File: docs/tutorials/gluon/learning_rate_finder.md (hunk 1 above)
[the remark this answers is not in the document]
Comment (ThomasDelteil) [reply docs/tutorials/gluon/learning_rate_finder.md]: When running inside the docker container the tutorial appears to sometimes hang when using multiple workers.

File: docs/tutorials/onnx/export_mxnet_to_onnx.md (hunk 3 above)
[the remark this answers is not in the document]
Comment (ThomasDelteil) [reply docs/tutorials/onnx/export_mxnet_to_onnx.md]: When putting some code inside a ` ```python` block, this will automatically get converted into a cell in the notebook. Currently this tutorial cannot be executed when downloaded from the website, because the help instructions for this function are treated as python code. I also un-pythoned this code block to avoid printing twice the same content when downloading the notebook.

File: docs/tutorials/gluon/learning_rate_schedules.md (hunk 2 above)
Comment (ThomasDelteil) [reply docs/tutorials/gluon/learning_rate_schedules.md]: @vishaalkapoor having too many workers for your DataLoader slows down your processing. Ideally your workers are in `[0, cpu_count - 2 - n_gpu]`
You want to use 0 workers when your data is already in memory and pre-processed, for example an ArrayDataset that contains your data pre-transformed and ready to be fed to GPU.

**Comment by ThomasDelteil** [comment]
@vishaalkapoor @anirudhacharya updated
@marcoabreu can you please advise how to set up a test job.

**Comment by marcoabreu** [comment]
Sure, please work with @lebeg. He will assist you this week

**Comment by ThomasDelteil** [comment]
@lebeg hey Anton can you please advise on how to set up a test job for that PR ? Thanks!

**Comment by anirudhacharya** [comment]
LGTM. Thanks @ThomasDelteil

**Comment by lebeg** [comment]
@ThomasDelteil Here you go:

http://jenkins.mxnet-ci-dev.amazon-ml.com/job/test-nightly-tutorials/view/change-requests/job/PR-13099/1/console

```
org.codehaus.groovy.control.MultipleCompilationErrorsException: startup failed:
WorkflowScript: 94: expecting '}', found ':' @ line 94, column 24.
       'Tutorial: Python2': {
                          ^
1 error
```

**Comment by lebeg** [comment]
Hm, I thought it will retry by itself... Anyway:
http://jenkins.mxnet-ci-dev.amazon-ml.com/job/test-nightly-tutorials/view/change-requests/job/PR-13099/2/

**Comment by ThomasDelteil** [comment]
Thanks @lebeg, I've pushed a new commit to remove the coverage analysis since it is not relevant and crashed the previous PR. I'll wait to see if it automatically triggers a new nightly build!

**Comment by marcoabreu** [comment]
I'd appreciate it if we could leave the coverage report in and instead work on getting it to work. Test coverage is an important metric and it would be good if we could track it for tutorials.

**Comment by ThomasDelteil** [comment]
It doesn't make sense for the tutorials since they run inside a subprocess created by the python process that is analysed by the code coverage. I don't see a way to enable test coverage inside the jupyter notebook executor and I don't think that is useful anyway, is a tutorial better because it covers more APIs? Should we trigger an alarm if we update a tutorial and it covers less of MXNet APIs? I don't think that's where the value of the tutorials are.

Straight dope tutorials do not have code coverage either for the same reasons.

**Comment by ThomasDelteil** [comment]
@marcoabreu @lebeg can you trigger the nightly job please?
Is there someone on Pacific time that can trigger such job? Thanks!

**Comment by marcoabreu** [comment]
I think the usefulness is that we can tag test coverage metrics for their source. This allows us to determine how much of our code is covered by tutorials and thus determine functions gaps (for example, we could see whether we are missing tutorials for quantization or the profiler API). But I agree, the intention of a tutorial is not to increase test coverage but I'd rather go from a user perspective and say that this number would express how many use-cases are covered by our tutorials.

**Comment by marcoabreu** [comment]
I have granted you permission to trigger jobs

**Comment by ThomasDelteil** [comment]
@marcoabreu thanks, I take back what I said and I agree it could indeed be interesting to leverage that data for tracking the coverage of the APIs for the tutorials, as an additional data point to help decide the content of future tutorials. 

One way I think we could do that would be to create a new suite of tests that convert the notebooks to python scripts and execute that instead of the notebook. 

However I will leave it as a future enhancement outside of the scope of re-enabling nightly jupyter notebook tests of the tutorials, which are focused on making sure the notebooks downloaded by the users from the website are running successfully in Python2 and Python3.

@marcoabreu looks like the two new suite passed, can you please review and merge ?

**Comment by marcoabreu** [comment]
That's a great idea. Would you mind creating an issue so we don't forget about it? :)

File: tests/tutorials/test_tutorials.py [hunk 5]
```diff
@@ -189,6 +189,4 @@ def test_vision_cnn_visualization():
 
 def test_control_flow():
     assert _test_tutorial_nb('control_flow/ControlFlowTutorial')
-
-def test_subgraphapi():
-    assert _test_tutorial_nb('c++/subgraphAPI')
```
Comment (marcoabreu) [inline tests/tutorials/test_tutorials.py]: Is there an issue that's tracking this? The subgraph API is something fairly new and users might be interested using it. Thus, we should make sure that our tutorials are being validated. Sorry if the tutorial is not testable - I didn't have a look.

Comment (ThomasDelteil) [reply tests/tutorials/test_tutorials.py]: It is a C++ tutorial that is not testable in Jupyter

**Comment by ThomasDelteil** [comment]
@marcoabreu @vishaalkapoor and @anirudhacharya approved the changes to the tutorials, you approved the CI part, I think we are good. Will create an issue for adding coverage of tutorials.
